src/service/fixed-layer.js

```diff
@@ -0,0 +1,429 @@
+          }
+        });
+      }
+    }).catch(error => {
+      // Fail silently.
+      setTimeout(() => {throw error;});
+    });
+  }
+
+  /**
+   * We currently only allow elements with `top: 0` or `bottom: 0`.
+   * @param {string} s
+   * @return {boolean}
+   */
+  isAllowedCoord_(s) {
+    return (!!s && parseInt(s, 10) == 0);
+  }
+
+  /**
+   * This method records the potentially fixed element. One of a more critical
+   * function - it records all selectors that may apply "fixed" to this element
+   * to check them later.
+   *
+   * @param {!Element} element
+   * @param {string} selector
+   * @param {!Object<string, !FixedElementDef>} idMap
+   * @private
+   */
+  setupFixedElement_(element, selector, idMap) {
+    let fixedId = element.getAttribute('i-amp-fixedid');
+    let fe;
+    if (fixedId) {
+      // Already seen.
+      fe = idMap[fixedId];
+      fe.selectors.push(selector);
+    } else {
+      fixedId = 'F' + (this.counter_++);
+      element.setAttribute('i-amp-fixedid', fixedId);
+      fe = {
+        id: fixedId,
+        element: element,
+        selectors: [selector],
+      };
+      this.fixedElements_.push(fe);
+      idMap[fixedId] = fe;
+    }
+  }
+
+  /**
+   * Mutates the fixed element. At this point it's determined that the element
+   * is indeed fixed. There are two main functions here:
+   *  1. `top` has to be updated to reflect viewer's paddingTop.
+   *  2. The element may need to be transfered to the separate fixed layer.
+   *
+   * @param {!FixedElementDef} fe
+   * @param {number} index
+   * @param {!FixedElementStateDef} state
+   * @private
+   */
+  mutateFixedElement_(fe, index, state) {
+    const element = fe.element;
+    const oldFixed = fe.fixedNow;
+    if (oldFixed == state.fixed) {
+      return;
+    }
+
+    fe.fixedNow = state.fixed;
+    if (state.fixed) {
+      // Update `top`. This is necessary to adjust position to the viewer's
+      // paddingTop.
+      if (state.top) {
+        element.style.top = `calc(${state.top} + ${this.paddingTop_}px)`;
+      }
+
+      // Move element to the fixed layer.
+      if (this.transfer_) {
+        if (state.transferrable) {
+          this.transferToFixedLayer_(fe, index, state);
+        } else {
+          this.returnFromFixedLayer_(fe);
+        }
+      }
+    } else {
+      // Reset `top` which was assigned above.
+      if (element.style.top) {
+        element.style.top = '';
+      }
+
+      // Move back to the BODY layer and reset transfer z-index.
+      this.returnFromFixedLayer_(fe);
+    }
+  }
+
+  /**
+   * @param {!FixedElementDef} fe
+   * @param {number} index
+   * @param {!FixedElementStateDef} state
+   * @private
+   */
+  transferToFixedLayer_(fe, index, state) {
+    const element = fe.element;
+    if (element.parentElement == this.fixedLayer_) {
+      return;
+    }
+
+    log.fine(TAG, 'transfer to fixed:', fe.id, fe.element);
+    console./*OK*/warn('transfer to fixed layer:', fe.element);
```

Comment:
Make this:
In order to improve scrolling performance in Safari, we now move the element to a fixed positioning layer.